### PR TITLE
kvserverbase: remove NonDeterministicError

### PR DESCRIPTION
**Pull request by tbg**
Reflecting on a recent change[^1] I realized we didn't need
`NonDeterministicError` any more since we have since fully embraced
`cockroachdb/errors` which provides redactable errors, so the "safe
explanation" functionality of this error implementation is unnecessary.

Furthermore, unexpected errors during raft application errors are by
default considered nondeterministic - it would be an invitation for
problems if this were otherwise. So contrary to what the name suggests,
using a `NonDeterministicError` never meant that we were doing anything
special to treat it as nondeterministic; there is really just one error
that gets special treatment, and it's in the other direction:
`apply.ErrRemoved` indicates that the replica got destroyed as a result
of command application, and is part of the regular Replica lifecycle.

Long story short, there wasn't a reason to keep `NonDeterministicError`
around, and this commit removes it.

[^1]: https://github.com/cockroachdb/cockroach/pull/91750/commits/a1d7605cc9823d38284cf447bdf02ea74249539d

Epic: CRDB-220
Release note: None

### CODE REVIEW

File: pkg/kv/kvserver/replica_raft.go [hunk 1]
```diff
@@ -667,7 +666,7 @@ var noSnap IncomingSnapshot
 // non-sensitive cue as to what happened.
 func (r *Replica) handleRaftReady(
 	ctx context.Context, inSnap IncomingSnapshot,
-) (handleRaftReadyStats, string, error) {
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_raft.go:670]: Yay for getting rid of this `string`, I was recently wondering why it had been there.

File: pkg/kv/kvserver/replica_raft.go [hunk 2]
```diff
@@ -864,11 +859,11 @@ func (r *Replica) handleRaftReadyRaftMuLocked(
 	appTask.SetMaxBatchSize(r.store.TestingKnobs().MaxApplicationBatchSize)
 	defer appTask.Close()
 	if err := appTask.Decode(ctx, rd.CommittedEntries); err != nil {
-		return stats, fmt.Sprint(kvserverbase.GetRedactedNonDeterministicFailureExplanation(err)), err
+		return stats, err
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/replica_raft.go:862]: Do you want to wrap this in `errors` too, or it's already?
Same below.

File: pkg/kv/kvserver/store_raft.go [hunk 3]
```diff
@@ -457,10 +456,8 @@ func (s *Store) processRaftSnapshotRequest(
 		// be there based on `stats.snap.applied`	but it is a questionable use of
 		// stats and more susceptible to bugs.
 		typ = removePlaceholderDropped
-		var expl string
-		var err error
-		stats, expl, err = r.handleRaftReadyRaftMuLocked(ctx, inSnap)
-		maybeFatalOnRaftReadyErr(ctx, expl, err)
+		stats, err := r.handleRaftReadyRaftMuLocked(ctx, inSnap)
```
Comment (pavelkalinnikov) [inline pkg/kv/kvserver/store_raft.go:459]: Nice.

**Review by pav-kv — APPROVED** [review]

**Comment by tbg** [comment]
bors r=pavelkalinnikov
